cadquery/assembly.py

```diff
@@ -434,32 +353,40 @@ def solve(self) -> "Assembly":
         Solve the constraints.
         """
 
-        # get all entities and number them
+        # Get all entities and number them. First entity is marked as locked
         ents = {}
 
         i = 0
-        lock_ix = 0
+        locked = []
         for c in self.constraints:
             for name in c.objects:
                 if name not in ents:
                     ents[name] = i
-                    if name == self.name:
-                        lock_ix = i
+                    if c.kind == "Fixed" or name == self.name:
+                        locked.append(i)
                     i += 1
 
+        # Lock the first occuring entity if needed.
+        if not locked:
+            locked.append(0)
```

Comment:
```suggestion
            unary_objects = [c.objects[0] for c in self.constraints if instance_of(c.kind, UnaryConstraintKind)]
            binary_objects = [c.objects[0] for c in self.constraints if instance_of(c.kind, BinaryConstraintKind)]
            for b in binary_objects: 
                if b not in unary_objects:
                    locked.append(ents[b])
                    break
```
When there are no explicit "Fixed" constraints, the current logic will lock the first occurring entity.  

Now with the new unary constraints this can be a problem if the first entity is part of a unary constraint.  I think it's important to avoid locking an entity that is part of a unary constraint as it will guarantee non-convergence of the optimizer. 

Something like the above changes might be implemented to lock only the first object of a binary constraint that is also not part of any unary constraint.